### ISSUE TITLE
md5sum failure results in packages not being installed

### ISSUE THREAD

**Issue by jfritz** [transcode-open/apt-cyg#37]
Tried doing a fresh install in cygwin x64, I'm seeing this behavior:

```bash
[Jeff@Jeff-PC3 ~]$ apt-cyg install wget
Installing wget
md5sum: standard input: no properly formatted MD5 checksum lines found
wget is not installed, using lynx as fallback
md5sum: standard input: no properly formatted MD5 checksum lines found
[Jeff@Jeff-PC3 ~]$ which wget
which: no wget in (/home/Jeff/bin:/usr/local/bin:/usr/bin:/cygdrive/c/Program Files (x86)/NVIDIA Corporation/PhysX/Common:/cygdrive/c/Windows/system32:/cygdrive/c/Windows:/cygdrive/c/Windows/System32/Wbem:/cygdrive/c/Windows/System32/WindowsPowerShell/v1.0:/sbin)
[Jeff@Jeff-PC3 ~]$
```

Will try playing around some more to get some more troubleshooting information.

**Comment by F3TTT** [comment]
I'm having the same issue.   They mention it here and say a reinstall should fix it but I'm having no such luck:  https://groups.google.com/forum/#!topic/apt-cyg-discuss/Dc-DYEInlP8

**Comment by jfritz** [comment]
Interesting. I'm using apt-cyg version 0.59 and have the issue. I tried reinstalling apt-cyg and no luck either.

**Comment by F3TTT** [comment]
ditto, I upgraded to 0.59 but no luck.

**Comment by F3TTT** [comment]
I don't get the: md5sum: standard input: no properly formatted MD5 checksum lines found

instead I get this(although I'm trying a different package since wget is already installed on this system):

$ apt-cyg install rtorrent
Working directory is /setup
Mirror is http://mirrors.kernel.org/sourceware/cygwin
--2015-03-24 08:11:23--  http://mirrors.kernel.org/sourceware/cygwin/x86_64/setup.bz2
Resolving mirrors.kernel.org (mirrors.kernel.org)... 149.20.37.36, 198.145.20.143, 2620:3:c000:a:0:1994:3:14, ...
Connecting to mirrors.kernel.org (mirrors.kernel.org)|149.20.37.36|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1269710 (1.2M) [application/octet-stream]
Saving to: ‘setup.bz2’

setup.bz2           100%[=====================>]   1.21M  1.65MB/s   in 0.7s

2015-03-24 08:11:24 (1.65 MB/s) - ‘setup.bz2’ saved [1269710/1269710]

Updated setup.ini

Installing rtorrent
Found package rtorrent
File ‘rtorrent-0.9.4-1.tar.xz’ already there; not retrieving.

MD5 sum did not match, exiting

ddoughty@TS001-EQLT769 ~
$

**Comment by jfritz** [comment]
Dan, I get the same error on a different (work) machine. 64-bit cygwin, 0.59 apt-cyg. I could swear almost 100% that this particular installation was working previously.

At home I reverted to 0.58 apt-cyg but didn't have time to test. Will report back later.

**Comment by jfritz** [comment]
The problem is that I don't think setup.ini uses md5 anymore, here is an example from the tree package:
```
@ tree
sdesc: "Display graphical directory tree"
ldesc: "Display an indented directory tree, using the same color assignments
as ls, via the LS_COLORS environment variable."
category: Utils
requires: cygwin
version: 1.7.0-1
install: x86_64/release/tree/tree-1.7.0-1.tar.xz 46456 3c777e777b2ccf9c5ad7b61798153014bcd05b097fbca384390bdfce5a71f0a46ce3cf151eb7a7ee2b3960e3959436ddbf1c53f7c3ed64a770e65363bc062a95
source: x86_64/release/tree/tree-1.7.0-1-src.tar.xz 116576 3f74348bfec6734a1a2cb1880ed950f3a80ff714c25f4d41d0a9d37f4ee9946faaad36ad7c2f77b0a46478551e26f75b119002c942caf0b2258f1ddf368d5b48
```

I think it's sha512:
```
[jsfritz@host tree]$ sha512sum.exe tree-1.7.0-1.tar.xz
3c777e777b2ccf9c5ad7b61798153014bcd05b097fbca384390bdfce5a71f0a46ce3cf151eb7a7ee2b3960e3959436ddbf1c53f7c3ed64a770e65363bc062a95 *tree-1.7.0-1.tar.xz
```

If you change line 320 (apt-cyg 0.59) from md5sum to sha512sum it works for me (see below). Per this mailing list message it appears they intended to change this because md5 is broken: https://sourceware.org/ml/cygwin/2015-03/msg00119.html 
```
[jsfritz@host ~]$ ./apt-cyg install tree

Installing tree
File ‘tree-1.7.0-1.tar.xz’ already there; not retrieving.

Unpacking...
Package tree requires the following packages, installing:
cygwin
Package cygwin is already installed, skipping
```

**Comment by F3TTT** [comment]
Confirmed, that worked for me as well.  Thanks for digging into it.

**Comment by mosbth** [comment]
Confirmed, the source for the pull-request #38 from @jfritz works for me.

**Comment by dakky** [comment]
Same here. thx for that fix

**Comment by ginister** [comment]
Thanks for the fix - Surprisingly lazy of apt-cyg developers! Has anybody come up with a solution that incorporates MD5 as well as SHA (In the case that MD5 checksums are still floating around)? I'd take a stab at it but I have no idea what the code is doing, and I don't want to break everything :)

**Comment by svnpenn** [comment]
> Surprisingly lazy of apt-cyg developers

@ginister I am the sole developer, and I recently lost my job, so I have been
focused on job hunting rather than this. More germane, Cygwin has been using MD5
for at least 10 years, and only recently switched to SHA this year.

If you are not happy with my work ethic, I will happily consider pull request
from you. If you are unwilling/unable to do this, then perhaps keep your
comments to yourself.

**Comment by benklaasen** [comment]
Hear hear. I find it difficult to understand the rudeness of people who get
amazing, software libre tools for free, find an issue and then instead of
attempting to contribute by fixing the issue, make ad hominem attacks on
the creators of these tools.

**Comment by CMCDragonkai** [comment]
Thanks, was very frustrated. But cygwin-ports still works with md5. So does this fix break cygwin-ports?

**Comment by johnjelinek** [comment]
@jfritz :+1: good find

**Comment by nwgat** [comment]
just installed apt-cyg and still broken :/

**Comment by jfritz** [comment]
Yes, looks like my PR #38 only addressed the main cygwin setup.ini change to sha256, but not cygwin-ports' continued use of md5. 

I don't use cygwin-ports, so I'm not familiar with the project. Someone who uses cygwin-ports should submit a new apt-cyg PR that works for both mainline cygwin and cygwin-ports.

Besides that, it's arguable that cygwin-ports should move to sha256. md5 is considered unsafe to use.

**Comment by svnpenn** [comment]
As can be seen here
http://mirrors.kernel.org/sourceware/cygwinports/x86_64/setup.bz2

Cygwin Ports is still using MD5. I am not going to accept a PR that fixes
something and breaks something else. So people can either download @jfritz
version or wait for me to fix it proper.

**Comment by svnpenn** [comment]
As can be seen here
http://mirrors.kernel.org/sourceware/cygwinports/x86_64/setup.bz2

Cygwin Ports is still using MD5. I am not going to accept a PR that fixes
something and breaks something else. So people can either download @jfritz
version or wait for someone to fix it proper.

**Comment by kou1okada** [comment]
My hack solve your problem.
https://github.com/kou1okada/apt-cyg/commit/cd55c387056f46c4d2823367fc25caf1c122cbc9

**Comment by kou1okada** [comment]
I wrote a pull request https://github.com/transcode-open/apt-cyg/pull/43

**Comment by svnpenn** [comment]
We will follow this issue to #43

**Comment by Fmajor** [comment]
i made this change in /bin/apt-cyg

    digactual1=`sha512sum $file | $GAWK '{print $1}'`
    digactual2=`md5sum $file | $GAWK '{print $1}'`
    if [ $digest != $digactual1 ] && [ $digest != $digactual2 ]; then

because i find that 
when i install python, it use sha512sum
but when i install ipython, it use md5sum.

wish it help

**Comment by kou1okada** [comment]
Why don't you update the latest version?

**Comment by Fmajor** [comment]
I was using the MobaXterm, which has a 0.57 version of apt-cyg
when i update to the latest, it report errors like this

     apt-cyg update

     awk: /etc/setup/setup.rc: No such file or directory
     awk: /etc/setup/setup.rc: No such file or directory
     mkdir: can't create directory '//x86': Read-only file system
     /bin/apt-cyg: line 149: cd: //x86: No such file or directory
     /x86/setup.bz2: Scheme missing.
     Error updating setup.ini, reverting

**Comment by kou1okada** [comment]
Hmm, I checked apt-cyg of MobaXterm bundled version, and I found that it is not the original version of apt-cyg.
They made some modifications for specializing to MobaXterm on an old original version of apt-cyg.
Also they sell a billing version of MobaXterm.
Thus, I make a suggestion that you ask for help first to MobaXterm distributor.
Maybe they update apt-cyg to the latest version and make some modifications for speializing to MobaXterm.

**Comment by Fmajor** [comment]
thank you for your suggestion.
I tried to get the python conding environment in MobaXterm, but finally i give up, there were so many problems in the cygwin environment....
maybe i consider to turn to the Virtual Machine...

**Comment by kou1okada** [comment]
Hmm, I understand why they need to modifications for specializing to MobaXterm.

    ➤ sha512sum --help
    BusyBox v1.22.1 (2014-09-26 07:33:17 CEST) multi-call binary.
     
    Usage: sha512sum [-c[sw]] [FILE]...
     
    Print or check SHA512 checksums
     
            -c      Check sums against list in FILEs
            -s      Don't output anything, status code shows success
            -w      Warn about improperly formatted checksum lines

Oh, it is not a standard command, it is a busybox.

**Comment by Fmajor** [comment]
hmm....
after the changes i made
    digactual1=`sha512sum $file | $GAWK '{print $1}'`
    ......
the install process can run without error
but i just CAN　ＮＯＴ　 run ipython, even git

**Comment by svnpenn** [comment]
@Fmajor `/etc/setup/setup.rc` is created by `setup-x86_64.exe`. So if you do not have it, you fucked up somewhere

**Comment by kou1okada** [comment]
I confirmed that mobaXterm uses cygwin, but it does not use `setup-x86.exe`.
So it does not have `/etc/setup` directory.
And large part of commands are a busybox.

**Comment by svnpenn** [comment]
I am not interested in supporting MobaXterm. If someone wants to submit a patch I will consider it.

**Comment by Fmajor** [comment]
yes, MobaXterm just use a little part of cygwin and do not update it frequently.
I have give up the thought to use it as a good-supported linux-like enviroment

**Comment by mredaelli** [comment]
I had the same problem with MobaXterm, but the new version 7.7 fixed it

**Comment by BagpipesJohnson** [comment]
Hey--Just downloaded by following instructions here:

http://stackoverflow.com/questions/17825023/installing-additional-packages-for-cygwin 

I am having the MD5 issue.

**Comment by caesargus** [comment]
Thank you @Fmajor - One of your comments assisted me in fixing my installation.

**Comment by restl3ss** [comment]
There's a very simple solution that doesn't break anything (though it's arguably more dangerous). 
Open apt-cyg with a text editor of your choice (that won't corrupt anything, so no notepad. notepad++ works) hit ctrl+f and type md5sum. Hit enter and it'll bring you to where the md5sum is verified. Simply comment out / delete the line that says "exit 1" and save. Now when you run apt-cyg you'll still get the warning about md5sum not matching, but the install will go through anyway.

**Comment by svnpenn** [comment]
I have locked this issue, and I have removed all comments about MobaXterm. I
solved the original issue over 6 months ago. If anyone wish to discuss MobaXterm
they can do it in another issue.
